sys/hashes/sha3.c

```diff
@@ -0,0 +1,457 @@
+ *                              @a delimitedSuffix must be 0x1F.
+ *                           - If the 7-bit sequence 1,1,0,1,0,0,0 is to be absorbed,
+ *                              @a delimitedSuffix must be 0x8B.
+ * @param  output          Pointer to the buffer where to store the output.
+ * @param  outputByteLen   The number of output bytes desired.
+ * @pre    One must have r+c=1600 and the rate a multiple of 8 bits in this implementation.
+ */
+static void Keccak(unsigned int rate, unsigned int capacity, const unsigned char *input,
+                   unsigned long long int inputByteLen, unsigned char delimitedSuffix, unsigned char *output,
+                   unsigned long long int outputByteLen);
+
+/**
+ *  Function to compute SHA3-256 on the input message. The output length is fixed to 32 bytes.
+ */
+static void FIPS202_SHA3_256(const unsigned char *input, unsigned int inputByteLen, unsigned char *output)
+{
+    Keccak(1088, 512, input, inputByteLen, 0x06, output, SHA3_256_DIGEST_LENGTH);
+}
+
+void sha3_256(void *digest, const void *data, size_t len)
+{
+    FIPS202_SHA3_256(data, len, digest);
+}
+
+void sha3_256_init(keccak_state_t *ctx)
+{
+    Keccak_init(ctx, 1088, 512, 0x06);
+}
+
+void sha3_update(keccak_state_t *ctx, const void *data, size_t len)
+{
+    Keccak_update(ctx, data, len);
+}
+
+void sha3_256_final(keccak_state_t *ctx, void *digest)
+{
+    Keccak_final(ctx, digest, SHA3_256_DIGEST_LENGTH);
+}
+
+
+/**
+ *  Function to compute SHA3-384 on the input message. The output length is fixed to 48 bytes.
+ */
+static void FIPS202_SHA3_384(const unsigned char *input, unsigned int inputByteLen, unsigned char *output)
+{
+    Keccak(832, 768, input, inputByteLen, 0x06, output, 48);
+}
+
+void sha3_384(void *digest, const void *data, size_t len)
+{
+    FIPS202_SHA3_384(data, len, digest);
+}
+
+void sha3_384_init(keccak_state_t *ctx)
+{
+    Keccak_init(ctx, 832, 768, 0x06);
+}
+
+void sha3_384_final(keccak_state_t *ctx, void *digest)
+{
+    Keccak_final(ctx, digest, SHA3_384_DIGEST_LENGTH);
+}
+
+/**
+ *  Function to compute SHA3-512 on the input message. The output length is fixed to 64 bytes.
+ */
+static void FIPS202_SHA3_512(const unsigned char *input, unsigned int inputByteLen, unsigned char *output)
+{
+    Keccak(576, 1024, input, inputByteLen, 0x06, output, 64);
+}
+
+void sha3_512(void *digest, const void *data, size_t len)
+{
+    FIPS202_SHA3_512(data, len, digest);
+}
+
+void sha3_512_init(keccak_state_t *ctx)
+{
+    Keccak_init(ctx, 576, 1024, 0x06);
+}
+
+void sha3_512_final(keccak_state_t *ctx, void *digest)
+{
+    Keccak_final(ctx, digest, SHA3_512_DIGEST_LENGTH);
+}
+
+/*
+   ================================================================
+   Technicalities
+   ================================================================
+ */
+
+typedef uint8_t UINT8;
+typedef uint64_t UINT64;
+typedef UINT64 tKeccakLane;
+
+#ifndef LITTLE_ENDIAN
```

Comment:
Technically you can define this in this file so the user doesn't need to get active:

```C
#if __BYTE_ORDER__ == __ORDER__LITTLE_ENDIAN__
#define LITTLE_ENDIAN
#endif
```

Reply:
71cd9bd4f207a20c7f059bd40a34a74c6b49b425

Reply:
I'll wait for Murdock and squash afterwards.